test/e2e/gateway/gateway_universal.go

```diff
@@ -0,0 +1,190 @@
+package gateway
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/gruntwork-io/terratest/modules/shell"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	config_core "github.com/kumahq/kuma/pkg/config/core"
+	"github.com/kumahq/kuma/test/e2e/trafficroute/testutil"
+	. "github.com/kumahq/kuma/test/framework"
+)
+
+func GatewayOnUniversal() {
+	var cluster *UniversalCluster
+
+	EchoServerUniversal := func(name string) InstallFunc {
+		return func(cluster Cluster) error {
+			const service = "echo-service"
+			token, err := cluster.GetKuma().GenerateDpToken("default", service)
+			if err != nil {
+				return err
+			}
+
+			return TestServerUniversal(
+				name,
+				"default",
+				token,
+				WithArgs([]string{"echo", "--instance", "universal"}),
+				WithServiceName(service),
+			)(cluster)
+		}
+	}
+
+	// GatewayClientUniversal runs an empty container that will
+	// function as a client for a gateway.
+	GatewayClientUniversal := func(name string) InstallFunc {
+		return func(cluster Cluster) error {
+			return cluster.DeployApp(WithName(name), WithoutDataplane(), WithVerbose())
+		}
+	}
+
+	GatewayProxyUniversal := func(name string) InstallFunc {
+		return func(cluster Cluster) error {
+			token, err := cluster.GetKuma().GenerateDpToken("default", "edge-gateway")
+			if err != nil {
+				return err
+			}
+
+			dataplaneYaml := `
+type: Dataplane
+mesh: default
+name: {{ name }}
+networking:
+  address:  {{ address }}
+  gateway:
+    type: BUILTIN
+    tags:
+      kuma.io/service: edge-gateway
+`
+			return cluster.DeployApp(
+				WithKumactlFlow(),
+				WithName(name),
+				WithToken(token),
+				WithVerbose(),
+				WithYaml(dataplaneYaml),
+			)
+		}
+	}
+
+	BeforeEach(func() {
+		cluster = NewUniversalCluster(NewTestingT(), Kuma1, Silent)
+		Expect(cluster).ToNot(BeNil())
+
+		deployOpts := append(KumaUniversalDeployOpts, WithVerbose())
+
+		err := NewClusterSetup().
+			Install(Kuma(config_core.Standalone, deployOpts...)).
+			Install(GatewayClientUniversal("gateway-client")).
+			Install(EchoServerUniversal("echo-server")).
+			Install(GatewayProxyUniversal("gateway-proxy")).
+			Setup(cluster)
+
+		// OK, this is messed up. The makefile rule that builds the
+		// kuma-universal:latest image that is used for e2e tests
+		// always rebuilds Kuma with Gateway disabled. This means
+		// that by default, no Gateway tests will work, so we use the
+		// `WithKumactlFlow` option to detect the unsupported gateway
+		// type early (when kumactl creates the dataplane resource),
+		// and skip the remaining tests.
```

Comment:
How can we fix it? Can we always build kuma with Gateway for E2E tests?

Reply:
Yeh it's fixable, but a bit subtle as described in the comment, so I'd prefer to tackle it in a separate PR. One way to fix this is to copy the build binaries into the containers at runtime, rather than when building the image. I think that this would improve the e2e workflow as well.